flink/src/main/java/org/apache/iceberg/flink/connector/sink/FileWriter.java

```diff
@@ -0,0 +1,241 @@
+      return this;
+    }
+
+    public Builder withHadooopConfig(final org.apache.hadoop.conf.Configuration hadoopConfig) {
+      this.hadoopConfig = hadoopConfig;
+      return this;
+    }
+
+    public Builder withSpec(final PartitionSpec spec) {
+      this.spec = spec;
+      return this;
+    }
+
+    public Builder withMetrics(final IcebergWriterTaskMetrics icebergWriterTaskMetrics) {
+      this.icebergWriterTaskMetrics = icebergWriterTaskMetrics;
+      return this;
+    }
+
+    public Builder withSubtaskMetrics(final IcebergWriterSubtaskMetrics icebergWriterSubtaskMetrics) {
+      this.icebergWriterSubtaskMetrics = icebergWriterSubtaskMetrics;
+      return this;
+    }
+
+    public Builder withSchema(final Schema avroSchema) {
+      this.avroSchema = avroSchema;
+      return this;
+    }
+
+    public Builder withVttsTimestampField(final String vttsTimestampField) {
+      this.vttsTimestampField = vttsTimestampField;
+      return this;
+    }
+
+    public Builder withVttsTimestampUnit(final TimeUnit vttsTimestampUnit) {
+      this.vttsTimestampUnit = vttsTimestampUnit;
+      return this;
+    }
+
+    public FileWriter build() {
+      Preconditions.checkArgument(this.format != null, "File format is required");
+      Preconditions.checkArgument(this.path != null, "File path is required");
+      Preconditions.checkArgument(this.timeService != null, "ProcessingTimeService is required");
+      Preconditions.checkArgument(this.partitioner != null, "Partitioner is required");
+      Preconditions.checkArgument(this.appender != null, "File appender is required");
+      Preconditions.checkArgument(this.hadoopConfig != null, "Hadoop config is required");
+      Preconditions.checkArgument(this.spec != null, "Partition spec is required");
+      Preconditions.checkArgument(this.icebergWriterTaskMetrics != null, "icebergWriterTaskMetrics is required");
+      Preconditions.checkArgument(this.avroSchema != null, "avroSchema is required");
+      Preconditions.checkArgument(this.vttsTimestampField != null, "vttsTimestampField is required");
+      Preconditions.checkArgument(this.vttsTimestampUnit != null, "vttsTimestampUnit is required");
+      return new FileWriter(this);
+    }
+  }
+
+  private FileWriter(Builder builder) {
+    format = builder.format;
+    path = builder.path;
+    timeService = builder.timeService;
+    creationTime = timeService.getCurrentProcessingTime();
+    partitioner = builder.partitioner;
+    appender = builder.appender;
+    hadoopConfig = builder.hadoopConfig;
+    spec = builder.spec;
+    icebergWriterTaskMetrics = builder.icebergWriterTaskMetrics;
+    icebergWriterSubtaskMetrics = builder.icebergWriterSubtaskMetrics;
+    watermarkTimeExtractor = new WatermarkTimeExtractor(
+        builder.avroSchema, builder.vttsTimestampField, builder.vttsTimestampUnit);
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  public long write(IndexedRecord record) {
```

Comment:
This should be an Iceberg generic record instead. Alternatively, we could parameterize this class and use a convention like the MR input format: `useHiveRows` or `usePigTuples` in configuration to switch the accepted type.

Reply:
Addressed. Use Record (within Iceberg) as the input of write().
Will put the alternative solution as you sugguested into future enhancement.